Check if the product exists in the system before adding to comparing list.

adarshkhatri opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->
Check if the product exists in the system before adding to comparing list.

### Description (*)
Check if the product exists in the system before adding to comparing list. If the product doesn't exist, then we simply do nothing.

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#23071: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Use this url http:://www.domain.com/catalog/product_compare/index/items/ab
2. Use this url http:://www.domain.com/catalog/product_compare/index/items/1234567890
Make sure product ID doesn't exist in the system.

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

adarshkhatri committed:
Check if product exists in the system before adding to comparing list.

Check if the product exists in the system before adding to comparing list. If the product doesn't exist, then we simply do nothing.

adarshkhatri committed:
Update ListCompare.php

adarshkhatri committed:
Update ListCompare.php

adarshkhatri committed:
Update ListCompare.php

adarshkhatri commented:
@magento give me test instance

magento-engcom-team commented:
Hi @adarshkhatri. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @adarshkhatri, here is your [new Magento instance](https://pr-23073.instances.magento-community.engineering).
Admin access: [https://pr-23073.instances.magento-community.engineering/admin](https://pr-23073.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

dmytro-ch commented on app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php:
```diff
@@ -67,13 +75,16 @@ public function __construct(
         \Magento\Catalog\Model\ResourceModel\Product\Compare\Item $catalogProductCompareItem,
         \Magento\Customer\Model\Session $customerSession,
         \Magento\Customer\Model\Visitor $customerVisitor,
+        \Magento\Catalog\Model\ProductFactory $catalogFactory = null,
```
Please don't change the order of constructor parameters.
Add the new parameter after `$data` for backward compatibility.

adarshkhatri replied:
fixed.

dmytro-ch commented on app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php:
```diff
@@ -51,6 +51,13 @@ class ListCompare extends \Magento\Framework\DataObject
+     *
+     * @var \Magento\Catalog\Model\ProductFactory
+     */
+    protected $_catalogFactory;
```
Please add new properties with the `private` scope of visibility.
I would also recommend avoiding underscores at the beginning of the private/protected method's name since this is the legacy approach according to the PSR-2 Coding Style Guide:
```
Property names SHOULD NOT be prefixed with a single underscore to indicate protected or private visibility.
```

adarshkhatri replied:
changed to private and removed underscore.

dmytro-ch commented on app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php:
```diff
@@ -92,7 +103,10 @@ public function addProduct($product)
             $item->addProductData($product);
-            $item->save();
+            $productId = $item->getProductId();
+            if ($productId && $this->_catalogFactory->create()->load($productId)->getId()) {
```
Please use `\Magento\Catalog\Model\ResourceModel\Product::load` method instead of deprecated model's one.
Entities must not be responsible for their own loading.

adarshkhatri replied:
Is it safe to use `\Magento\Catalog\Api\ProductRepositoryInterface` then?

Sample:

Constructor
```
        array $data = [],
        \Magento\Catalog\Api\ProductRepositoryInterface $productRepository = null
$this->productRepository = $productRepository ?: \Magento\Framework\App\ObjectManager::getInstance()
            ->get(\Magento\Catalog\Api\ProductRepositoryInterface::class);

```

and call it like:
```
if ($productId && $this->productRepository->getById($productId)) {
                $item->save();
            }
```

orlangur replied:
@adarshkhatri yep, repository is actually more preferable compared to resource model.

adarshkhatri replied:
used Product Repo to load product.

dmytro-ch replied:
IMO it's better to use resource model from the performance perspective, since we just need to verify the entity exists :)
Nonetheless, it's not so critical.

adarshkhatri replied:
I ended up using search criteria builder. Please check my latest commit.

arnoudhgz commented:
@adarshkhatri @dmytro-ch Is it not better to check if the product exists as first thing in the `addProduct` method? It does not make sense to do anything else (like `_addVisitorToItem` and `loadByProduct`) in this method if the product does not exist.

adarshkhatri committed:
Update ListCompare.php

orlangur commented on app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php:
```diff
@@ -108,7 +118,9 @@ public function addProducts($productIds)
         if (is_array($productIds)) {
             foreach ($productIds as $productId) {
-                $this->addProduct($productId);
+                if ($this->productRepository->getById($productId)->getId()) {
```
This seems to be not effective. Please obtain all products at once with `SearchCriteria` or something.

adarshkhatri replied:
used SearchCriteriaBuilder

orlangur commented on app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php:
```diff
@@ -51,6 +51,13 @@ class ListCompare extends \Magento\Framework\DataObject
     protected $_compareItemFactory;
 
+    /**
+     * Catalog factory
```
Outdated comment, please remove.

orlangur reviewed (changes requested):
After changes are applied and all builds are green, please squash them into a single commit so that we have perfectly clean history 😉

adarshkhatri committed:
Update ListCompare.php

Using searchCriteriaBuilder to get the products from the system, filtered by (passed) productIds and then iterate that list of products to add to compare list.

adarshkhatri committed:
Update ListCompare.php

adarshkhatri commented:
@magento give me test instance

magento-engcom-team commented:
Hi @adarshkhatri. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @adarshkhatri, here is your [new Magento instance](https://pr-23073.instances.magento-community.engineering).
Admin access: [https://pr-23073.instances.magento-community.engineering/admin](https://pr-23073.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

orlangur commented on app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php:
```diff
@@ -51,6 +51,18 @@ class ListCompare extends \Magento\Framework\DataObject
     protected $_compareItemFactory;
 
+    /**
+     * Product Repository
```
Such comment is useless, please remove.

adarshkhatri replied:
removed.

orlangur commented on app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php:
```diff
@@ -60,20 +72,26 @@ class ListCompare extends \Magento\Framework\DataObject
-        array $data = []
+        array $data = [],
+        \Magento\Catalog\Api\ProductRepositoryInterface $productRepository,
+        \Magento\Framework\Api\SearchCriteriaBuilder $searchCriteriaBuilder
```
Inject in backward-compatible manner, see https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/

adarshkhatri replied:
fixed.

orlangur reviewed (changes requested):
After changes are applied and all builds are green, please squash them into a single commit so that we have perfectly clean history 😉

adarshkhatri committed:
Adding backward compatibility

Adding backward compatibility for constructor parameters.

adarshkhatri committed:
Fixing typo "_condProdCombineF"

Fixing typo "_condProdCombineF". Replacing it with "_condProdCombineFactory".

sidolov commented on app/code/Magento/SalesRule/Model/Rule.php:
```diff
@@ -160,7 +160,7 @@ class Rule extends \Magento\Rule\Model\AbstractModel
      * @var \Magento\SalesRule\Model\Rule\Condition\Product\CombineFactory
      */
-    protected $_condProdCombineF;
+    protected $_condProdCombineFactory;
```
Unfortunately, rename protected vars is backward incompatible according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/), please, revert

adarshkhatri committed:
Update Rule.php

adarshkhatri committed:
Only existing products to be added to compare list

Currently, the system tries to add any product id in to compare list, which is why it fails and throws an error.

Allow only existing products to be added to compare list.

adarshkhatri committed:
Merge pull request #1 from magento/2.3-develop

update to recent version

adarshkhatri committed:
Removing extra comment.

adarshkhatri commented:
@magento give me test instance

magento-engcom-team commented:
Hi @adarshkhatri. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @adarshkhatri, here is your [new Magento instance](https://pr-23073.instances.magento-community.engineering).
Admin access: [https://pr-23073.instances.magento-community.engineering/admin](https://pr-23073.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

adarshkhatri committed:
Removed php mess.

sidolov commented:
@adarshkhatri please, cover changes with automated tests

adarshkhatri commented:
> @adarshkhatri please, cover changes with automated tests

I Couldn't understand them. 

First fail:

> PHP Code Mess has found error(s):
> /var/www/html/app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php:20	The class Magento\Catalog\Model\Product\Compare\ListCompare uses sessions or cookies while not being a part of HTML Presentation layer
> 
> Failed asserting that 2 matches expected 0.


Second fail: 

> Database Compare / Database Compare build failing 12 days ago in 30m 38s
> No build reports available

sidolov commented:
@adarshkhatri test failures caused by legacy implementation and you don't need to fix them. Please, cover newly added logic by tests

adarshkhatri commented:
> @adarshkhatri test failures caused by legacy implementation and you don't need to fix them. Please, cover newly added logic by tests

Just couldn't understand what's the logic added by test in this?

Should I just add ` @SuppressWarnings(PHPMD.CookieAndSessionMisuse)` in the class?

Thanks.

sidolov commented:
@adarshkhatri usage of the repository was added to the method, it should be covered at least by unit test.

dmytro-ch commented:
Closing this PR due to the unnecessity of the mentioned improvement.
Feel free to reopen this PR if you wish to provide solid refactoring for undesirable approaches.
Thank you!
